src/ophyd_async/core/_device.py

```diff
@@ -60,23 +69,32 @@ def set_name(self, name: str):
         name:
             New name to set
         """
-
-        # Ensure self.log is recreated after a name change
-        if hasattr(self, "log"):
-            del self.log
-
         self._name = name
-        for attr_name, child in self.children():
-            child_name = f"{name}-{attr_name.rstrip('_')}" if name else ""
+        # Ensure self.log is recreated after a name change
+        self.log = LoggerAdapter(
+            getLogger("ophyd_async.devices"), {"ophyd_async_device_name": self.name}
+        )
+        for child_name, child in self.children():
+            child_name = f"{self.name}-{child_name.rstrip('_')}" if self.name else ""
             child.set_name(child_name)
-            child.parent = self
+
+    def __setattr__(self, name: str, value: Any) -> None:
+        if name == "parent":
+            if self.parent not in (value, None):
+                raise TypeError(
+                    f"Cannot set the parent of {self} to be {value}: "
```

Comment:
This is now breaking the Thawer as the ThawingTimer has a reference to the control signal of the parent Thawer Device.

Can we please not make assumptions about what subclasses may be doing with their own attributes? Or at the very least, allow them to have unmolested underscore private attribs?

Reply:
Done in e2df3248ea28619afc820080b12c755b5860f96c

Reply:
I may be having buyer's remorse on this but would like to solicit feedback from users of ophyd_async before asking to revert this change:

Previously, we had control over whether something was parented or not by deciding when to call the `super()._init__()` method, if the signal was present it got parented otherwise not. From conversation with @coretl apparently this was unintended behaviour but it seems we have been using it in dodal anyway.

With the change to parent on setting the attribute we lost control and all devices are parented - which resulted in the above change as it broke the Thawer as it has a device which has a private reference to a signal. However we also have some devices which are parented but had the signals private (underscore prefixed) because they aren't interesting to users of the device, and are only implementation, like the attenuator and the mirror voltages.

To some extent though private signals are still very visible anyway as they are exposed via events etc. so I don't know whether making them public is that big of a deal or not.

I think if this change were to be reverted though, we would ideally need some other explicit way of being able to control whether a signal is a child or not.

Reply:
> . From conversation with @coretl apparently this was unintended behaviour but it seems we have been using it in dodal anyway.

https://www.hyrumslaw.com/ strikes again

Reply:
Only just catching up on some of this, but as a simple "user" of ophyd-async, I agree it's surprising if `self.foo = some_signal(...)` and `self._foo = some_signal(...)` have different behaviour. That feels very magical & wrong.

Is there a more explicit API we could use?

```python
with self.dont_overwrite_parent():
    self.foo = some_signal_reference
```
?

Reply:
The other option is if we can find a way to make all this work with immutable attributes

Reply:
Discussed this with @rtuck99 and @callumforrester and decided that:
- Having different parent-child-naming behaviour for `self._foo` and `self.foo` is indeed surprising behaviour, so will revert the change so they behave the same
- To solve the `ThawerTimer` case we will provide a new box-like class, that is simply a container for a Device. This means it can be a public or private member of the class, but not take part in the parent-child-naming behaviour of Device, like we do with `DetectorWriter` and other logic classes:
```python
class Box(Generic[T]):
    def __init__(self, obj: T): ...
    def get(self) -> T: ...

class DeviceWithRefToSignal(Device):
    def __init__(self, signal: SignalRW[int[):
        self.signal_box = Box(signal)

    def set(self, value) -> AsyncStatus:
        signal = self.signal_box.get()  # typing knows that signal is a SignalRW[int]
        return signal.set(value + 1)
```

The questions:
- If `parent.name == "foo"`, and `parent._bar` is a `Device`, should `parent._bar.name` be "parent-bar" or "parent-_bar"?
- What should `Box` be called?

Reply:
I could model it after a [`weakref.proxy`](https://docs.python.org/3/library/weakref.html#weakref.proxy):
```python
class Reference(Generic[T]):
    def __init__(self, obj: T):
        self._obj = obj
    def __call__(self) -> T:
        return self._obj

class DeviceWithRefToSignal(Device):
    def __init__(self, signal: SignalRW[int[):
        self.signal_ref = Reference(signal)

    def set(self, value) -> AsyncStatus:
        return self.signal_ref().set(value + 1)
```
@Tom-Willemsen any better?

Reply:
That seeems fine to me at first glance - assuming the docstring / docs in general will explain why one might need to use it. Thanks!

Reply:
Sorry, fractionally too late to the party but only minor. Can we add a mention of `Reference` to this error message so that people know it is possible, they just have to do it in a different way?